x-pack/plugins/apm/server/lib/environments/get_all_environments.ts

```diff
@@ -48,6 +49,7 @@ export async function getAllEnvironments({
           terms: {
             field: SERVICE_ENVIRONMENT,
             size: 100,
+            ...(!serviceNameFilter.length ? { min_doc_count: 0 } : {}),
```

Comment:
Not a comment on the actual feasibility of this approach but a general comment about conditional logic in our ES queries.

I think it's useful to co-locate logic that is meant to go together. For instance, if we add this behaviour but choose to delete it again, it'll be much easier if it's co-located ([optimize for deletion](https://kellysutton.com/2017/05/29/deletability.html)).
Similarly, if there is an issue with this implementation and somebody (new) has to debug it, it's much easier if all the related lines are co-located so they don't have to study the es query in detail in case there are other places where we handle this behaviour. Additionally it's easier to write a meaningful comment that applies to all of the lines:

```js
// return results immediately for unbound queries (query without a filter context)
if (!serviceNameFilter) {
 params.body.timeout = '1ms';
 params.aggs.environments.terms.min_doc_count = 0;
}
```

Reply:
Agree that it's easier to understand. I'll need to look into whether TypeScript likes this though. I may need to look for something that has the same sentiment, but also has type-safety.

Reply:
@sqren this didn't work out for me, TypeScript started complaining. I don't feel it's worth it to add things like @ts-expect-error, so I'll just leave it as is. I did add a comment to explain what it does.

Reply:
Fair enough. I still thing it would have been worth co-locating it. It's not a big deal in a small query like this but it all adds up. I also don't like `@ts-expect-error` but I do like having a single coherent chunk of logic:

```ts
  if (!serviceNameFilter) {
    // @ts-expect-error
    params.body.timeout = '1ms';
    // @ts-expect-error
    params.aggs.environments.terms.min_doc_count = 0;
  }
```